amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/async/parser/AsyncMessageParser.scala

```diff
@@ -149,9 +150,18 @@ abstract class AsyncMessagePopulator()(implicit ctx: AsyncWebApiContext) extends
 
     map.key(
       "headers",
-      entry =>
-        parseHeaderSchema(entry, message.id) foreach { param =>
-          message.set(MessageModel.Headers, AmfArray(Seq(param), Annotations(entry.value)), Annotations(entry))
+      entry => {
+        AsyncApiTypeParser(entry, shape => shape.withName("schema").adopted(message.id), JSONSchemaDraft7SchemaVersion)
+          .parse()
+          .foreach {
+            case n: NodeShape =>
+              message.set(MessageModel.HeaderSchema, n, Annotations(entry))
+            case _ =>
+              ctx.eh.violation(ParserSideValidations.HeaderMustBeObject,
```

Comment:
[OOF] - A good improvement is to stop parsing the shape once we know its a nodeShape as we don't assign it the instance if it isn't a NodeShape. Probably not significant but still a processing saving